File: Core/Database/Domain/Core/Derivations/Default/Derivation.cs
// <copyright file="DomainDerivationCycle.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Database.Domain.Derivations.Default
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Database.Data;
    using Database.Derivations;
    using Object = Domain.Object;

    public class Derivation
    {
        public Derivation(ITransaction transaction, IValidation validation, Engine engine, int maxCycles)
        {
            this.Transaction = transaction;
            this.Validation = validation;
            this.Engine = engine;
            this.MaxCycles = maxCycles;
        }

        public ITransaction Transaction { get; }

        public IValidation Validation { get; }

        public Engine Engine { get; }

        public int MaxCycles { get; }

        public AccumulatedChangeSet Execute()
        {
            var domainCycles = 0;

            AccumulatedChangeSet domainAccumulatedChangeSet = null;

            domainAccumulatedChangeSet = new AccumulatedChangeSet();
            var domainValidation = new DomainValidation(this.Validation);

            var changeSet = this.Transaction.Checkpoint();
            domainAccumulatedChangeSet.Add(changeSet);

            while (changeSet.Associations.Any() || changeSet.Roles.Any() || changeSet.Created.Any() ||
                   changeSet.Deleted.Any())
            {
                if (++domainCycles > this.MaxCycles)
                {
                    throw new Exception("Maximum amount of domain derivation cycles detected");
                }

                // Initialization
                if (changeSet.Created.Any())
                {
                    var newObjects = changeSet.Created.Select(v => (Object)v.GetObject());
                    foreach (var newObject in newObjects)
                    {
                        newObject.OnInit();
                    }
                }

                var domainCycle = new DomainDerivationCycle
                {
                    ChangeSet = changeSet,
                    Transaction = this.Transaction,
                    Validation = domainValidation
                };

                var matchesByDerivation = new Dictionary<IRule, IEnumerable<IObject>>();
                foreach (var rule in this.Engine.ClassesByRule.Keys)
                {
                    var matches = new HashSet<IObject>();

                    foreach (var pattern in rule.Patterns)
                    {
                        var source = pattern switch
                        {

                            RolePattern { ObjectType: null } rolePattern => changeSet
                                .AssociationsByRoleType
                                .Where(v => v.Key.Equals(rolePattern.RoleType))
                                .SelectMany(v => this.Transaction.Instantiate(v.Value)),

                            RolePattern { ObjectType: { } } rolePattern => changeSet
                                .AssociationsByRoleType
                                .Where(v => v.Key.RelationType.Equals(rolePattern.RoleType.RelationType))
                                .SelectMany(v => this.Transaction.Instantiate(v.Value))
                                .Where(v => rolePattern.ObjectType.IsAssignableFrom(v.Strategy.Class)),

                            AssociationPattern associationPattern => changeSet
                                .RolesByAssociationType
                                .Where(v => v.Key.Equals(associationPattern.AssociationType))
                                .SelectMany(v => this.Transaction.Instantiate(v.Value)),

                            _ => Array.Empty<IObject>()
                        };

                        if (source != null)
                        {
                            if (pattern.Steps?.Length > 0)
                            {
                                var step = new Step(pattern.Steps);
                                source = source.SelectMany(v => step.Get(v));
                            }

                            if (pattern.OfType != null)
                            {
                                source = source.Where(v => pattern.OfType.IsAssignableFrom(v.Strategy.Class));
                            }

                            matches.UnionWith(source);
                        }
                    }

                    if (matches.Count > 0)
                    {
                        matchesByDerivation[rule] = matches;
                    }
                }

                // TODO: Prefetching

                foreach (var kvp in matchesByDerivation)
                {
                    var domainDerivation = kvp.Key;
                    var matches = kvp.Value;
                    domainDerivation.Derive(domainCycle, matches);
                }

                changeSet = this.Transaction.Checkpoint();
                domainAccumulatedChangeSet.Add(changeSet);
            }

            return domainAccumulatedChangeSet;
        }
    }
}

File: typescript/modules/libs/workspace/adapters/system/src/lib/session/derivation/derivation.ts
import { ICycle, IDerivation, IObject, IRule, ISession, IValidation, resolve } from '@allors/workspace/domain/system';
import { Engine } from './engine';

export class Derivation implements IDerivation {
  public constructor(public session: ISession, public engine: Engine, public activeRules: Set<IRule>, public maxCycles: number) {
    this.validation = { errors: [] };
  }

  validation: IValidation;

  public execute(): IValidation {
    let cycles = 0;
    let changeSet = this.session.checkpoint();

    while (changeSet.rolesByAssociationType?.size > 0 || changeSet.associationsByRoleType?.size > 0 || changeSet.created?.size > 0 || changeSet.instantiated?.size > 0) {
      if (++cycles > this.maxCycles) {
        throw new Error('Maximum amount of domain derivation cycles detected');
      }

      const cycle: ICycle = {
        changeSet,
        session: this.session,
        validation: this.validation,
      };

      const matchesByRule: Map<IRule, Set<IObject>> = new Map();

      if (changeSet.instantiated != null) {
        for (const instantiated of changeSet.instantiated.values()) {
          const cls = instantiated.strategy.cls;
          const rules = this.engine.rulesByClass.get(cls)?.filter((v) => this.activeRules.has(v));

          if (rules != null) {
            for (const rule of rules) {
              const patternsByClass = this.engine.patternsByClassByRule.get(rule);
              const patterns = patternsByClass?.get(cls);
              if (patterns != null) {
                for (const pattern of patterns) {
                  let matches = matchesByRule.get(rule);
                  if (matches == null) {
                    matches = new Set();
                    matchesByRule.set(rule, matches);
                  }

                  let source: IObject[];

                  if (pattern.tree != null) {
                    source = pattern.tree.reduce((acc, v) => {
                      for (const obj of resolve(instantiated, v, true)) {
                        acc.push(obj);
                      }
                      return acc;
                    }, []);
                  } else {
                    source = [instantiated];
                  }

                  if (pattern.ofType != null) {
                    source = source.filter((v) => pattern.ofType.isAssignableFrom(v.strategy.cls));
                  }

                  for (const obj of source) {
                    matches.add(obj);
                  }
                }
              }
            }
          }
        }
      }

      for (const [roleType, associations] of changeSet.associationsByRoleType) {
        for (const association of associations) {
          const cls = association.strategy.cls;
          const patterns = this.engine.patternsByRoleTypeByClass.get(cls)?.get(roleType);

          if (patterns != null) {
            for (const pattern of patterns) {
              const rule = this.engine.ruleByPattern.get(pattern);
              if (!this.activeRules.has(rule)) {
                continue;
              }

              let matches = matchesByRule.get(rule);
              if (matches == null) {
                matches = new Set();
                matchesByRule.set(rule, matches);
              }

              let source: IObject[];

              if (pattern.tree != null) {
                source = pattern.tree.reduce((acc, v) => {
                  acc.push(resolve(association, v, true));
                  return acc;
                }, []);
              } else {
                source = [association];
              }

              if (pattern.ofType != null) {
                source = source.filter((v) => pattern.ofType.isAssignableFrom(v.strategy.cls));
              }

              for (const obj of source) {
                matches.add(obj);
              }
            }
          }
        }
      }

      for (const [associationType, roles] of changeSet.rolesByAssociationType) {
        for (const association of roles) {
          const cls = association.strategy.cls;
          const patterns = this.engine.patternsByAssociationTypeByClass.get(cls)?.get(associationType);
          if (patterns != null) {
            for (const pattern of patterns) {
              const rule = this.engine.ruleByPattern.get(pattern);
              if (!this.activeRules.has(rule)) {
                continue;
              }

              let matches = matchesByRule.get(rule);
              if (matches == null) {
                matches = new Set();
                matchesByRule.set(rule, matches);
              }

              let source: IObject[];

              if (pattern.tree != null) {
                source = pattern.tree.reduce((acc, v) => {
                  acc.push(resolve(association, v, true));
                  return acc;
                }, []);
              } else {
                source = [association];
              }

              if (pattern.ofType != null) {
                source = source.filter((v) => pattern.ofType.isAssignableFrom(v.strategy.cls));
              }

              for (const obj of source) {
                matches.add(obj);
              }
            }
          }
        }
      }

      for (const [rule, matches] of matchesByRule) {
        rule.derive(cycle, [...matches]);
      }

      changeSet = this.session.checkpoint();
    }

    return this.validation;
  }
}
